### PR TITLE
[Content management] client POC

### PR DESCRIPTION
**Pull request by Dosant**
## Summary

built on top of [wip server POC](https://github.com/elastic/kibana/pull/148791) but with client cache layer  

The tech doc for the client-side content service: https://docs.google.com/document/d/13aOspEx_yGcdaRwXZTAOCaiPuAIZlcrkZopvR0cr2_k/edit#

The client POC consists of two independent parts:

- **Content Type Registry** – A client-side registry for registering a content type. Registrars provide presentational information about the content type such as display name, description, icon, etc.
- **Content Client** – A client-side service and set of helpers exposed to other plugins and our components for querying and mutating the content. Includes features such as cache, requests deduping, retries, etc.. The tech doc and the POC proposes to use react-query internally and for shaping an API

Untitled-2023-01-05-1448-3

### CODE REVIEW

File: src/plugins/content_management/public/registry/registry.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { ContentTypeDetails } from './types';
+import { ContentType } from './content_type';
+
+export class ContentRegistry {
```
Comment (loixlab) [inline src/plugins/content_management/public/registry/registry.ts:12]: I wonder why we need the registry client side. Couldn't we have this on the server and a method (cached) to fetch once those details from the server?

File: src/plugins/content_management/public/content_client/content_client.tsx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+ */
+
+import { QueryClient, QueryClientProvider, useMutation, useQuery } from '@tanstack/react-query';
+import React from 'react';
+import type { RpcClient } from '../rpc';
+import { createQueryObservable } from './query_observable';
+import type { SearchIn } from '../../common';
+import { CreateIn, CreateOut } from '../../common';
+
+const contentQueryKeyBuilder = {
```
Comment (loixlab) [inline src/plugins/content_management/public/content_client/content_client.tsx:16]: nit: Can we maybe shorten this to `queryKeyBuilder`?

File: src/plugins/content_management/public/content_client/content_client.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    type,
+    id,
+  }: {
+    type: string;
+    id: string;
+  }): Promise<ContentItem> {
+    return this.queryClient.fetchQuery(this.contentQueryOptionBuilder.get<ContentItem>(type, id));
+  }
+
+  get$<ContentItem extends object = object>({ type, id }: { type: string; id: string }) {
```
Comment (loixlab) [inline src/plugins/content_management/public/content_client/content_client.tsx:85]: I am not sure I fully understand what those observable version of the method do 😊

File: src/plugins/content_management/public/content_client/query_observable.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  queryOptions: QueryObserverOptions<TQueryFnData, TError, TData, TQueryData, TQueryKey>
+) => {
+  const queryObserver = new QueryObserver(
+    queryClient,
+    queryClient.defaultQueryOptions(queryOptions)
+  );
+
+  return new Observable<QueryObserverResult<TData, TError>>((subscriber) => {
+    const unsubscribe = queryObserver.subscribe(
+      // notifyManager is a singleton that batches updates across react query
```
Comment (loixlab) [inline src/plugins/content_management/public/content_client/query_observable.ts:36]: Strange I can't find anything about notifyManager in their docs... 🤔 

Screenshot 2023-01-30 at 13 52 49

File: src/plugins/content_management/public/plugin.ts [hunk 5]
```diff
@@ -11,22 +11,37 @@ import { ManagementAppMountParams, ManagementSetup } from '@kbn/management-plugi
...
+import { ContentManagementPublicStart, ContentManagementPublicSetup } from './types';
+import { ContentClient } from './content_client';
+import { ContentRegistry } from './registry';
 
 interface SetupDependencies {
   management: ManagementSetup;
 }
 
 export class ContentManagementPlugin implements Plugin {
   private rpcClient: RpcClient | undefined;
```
Comment (loixlab) [inline src/plugins/content_management/public/plugin.ts:23]: We probably don't need this private property anymore

File: src/plugins/content_management/public/content_client/content_client.tsx (hunk 3 above)
Comment (Dosant) [reply src/plugins/content_management/public/content_client/content_client.tsx:85]: Have you seen https://docs.google.com/document/d/13aOspEx_yGcdaRwXZTAOCaiPuAIZlcrkZopvR0cr2_k/edit#heading=h.oijoa468kc4q ? 

The idea is that cache is reactive, so the main way to access data would be through the react hook or RXJS observable. Regular promise based method can be used in the cases where reactivity is not needed

File: src/plugins/content_management/public/content_client/query_observable.ts (hunk 4 above)
Comment (Dosant) [reply src/plugins/content_management/public/content_client/query_observable.ts:36]: I don't think it is a public api. It is exported from query-core and is used by react (or other libs) implementations. I learned about it from the useQuery hook source

File: src/plugins/content_management/public/registry/registry.ts (hunk 1 above)
Comment (Dosant) [reply src/plugins/content_management/public/registry/registry.ts:12]: Good question. I'll put this question here https://docs.google.com/document/d/13aOspEx_yGcdaRwXZTAOCaiPuAIZlcrkZopvR0cr2_k/edit# and will iterate on it

File: src/plugins/content_management/public/content_client/query_observable.ts (hunk 4 above)
Comment (loixlab) [reply src/plugins/content_management/public/content_client/query_observable.ts:36]: Ok. Isn't it risky to not use a public api?

File: src/plugins/content_management/public/content_client/content_client.tsx (hunk 3 above)
Comment (loixlab) [reply src/plugins/content_management/public/content_client/content_client.tsx:85]: Ah I see, indeed this is outside React context so we need a way to make it reactive. 👍

**Comment by TinaHeiligers** [comment]
cc @TinaHeiligers @jloleysens
